### PR TITLE
test: Add SG and OIDC Clean-up

### PR DESCRIPTION
**Pull request by engedaam**
Fixes #N/A 

**Description**
- Add clean up of SG and leaked OICD 

**How was this change tested?**
- Manually tested 

**Does this change impact docs?**
- [] Yes, PR includes docs updates
- [] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/sweeper.yaml [hunk 1]
```diff
@@ -9,14 +9,18 @@ permissions:
 jobs:
   sweeper:
     if: github.repository == 'aws/karpenter' || github.event_name == 'workflow_dispatch'
+    strategy:
```
Comment (jonathan-innis) [inline .github/workflows/sweeper.yaml:12]: Nice 🎉

File: test/hack/cleanup/main.go [hunk 2]
```diff
@@ -105,6 +132,22 @@ func main() {
 	if err := fireMetric(ctx, cloudWatchClient, "LaunchTemplatesDeleted", float64(deleted)); err != nil {
 		logger.With("name", "LaunchTemplatesDeleted").Errorf("firing metric, %v", err)
 	}
+
+	names = getOldOICD(ctx, iamClient, expirationTime)
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:136]:
```suggestion
	names = getOldOIDCProviders(ctx, iamClient, expirationTime)
```

File: test/hack/cleanup/main.go [hunk 3]
```diff
@@ -71,10 +76,32 @@ func main() {
 
 	// Terminate any old stacks that were provisioned as part of testing
 	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
+	sgInAccount := lo.Must(ec2Client.DescribeSecurityGroups(ctx, &ec2.DescribeSecurityGroupsInput{
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:79]: Can we just delete these security groups if they are beyond the expiration time like the other things?

Comment (jonathan-innis) [inline test/hack/cleanup/main.go:79]: Also, a follow-up question: is this just a back-stop for the security group not deleting as part of the actual test run? Can we make changes so that, in the general case, we delete the security group and successfully delete the VPC but we only use the sweeper if we fail to do so for some other reason

File: test/hack/cleanup/main.go [hunk 4]
```diff
@@ -208,3 +251,23 @@ func getOldLaunchTemplates(ctx context.Context, ec2Client *ec2.Client, expiratio
 	}
 	return names
 }
+
+func getOldOICD(ctx context.Context, iamClient *iam.Client, expirationTime time.Time) (names []string) {
+	testSuite := []string{"upgrade", "chaos", "consolidation", "drift", "integration", "interruption", "ipv6", "machine", "scale", "utilization"}
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:256]: Is it possible for us to identify this OIDC provider by some other tag that is only added by a test run? Otherwise, we are going to have to keep expanding this list as we add tests and we are going to forget to maintain this list

File: test/hack/cleanup/main.go [hunk 5]
```diff
@@ -105,6 +132,22 @@ func main() {
 	if err := fireMetric(ctx, cloudWatchClient, "LaunchTemplatesDeleted", float64(deleted)); err != nil {
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:132]: Side task: can we configure these metrics to fire to timestream, since we are using this for our metrics storage now?

File: test/hack/cleanup/main.go [hunk 6]
```diff
@@ -105,6 +132,22 @@ func main() {
...
 		logger.With("name", "LaunchTemplatesDeleted").Errorf("firing metric, %v", err)
 	}
+
+	names = getOldOICD(ctx, iamClient, expirationTime)
+	deleted = 0
+	for i := range names {
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:138]:
```suggestion
	for i := range arns {
```
If this is ARNs, can we name the variable as such?

File: test/hack/cleanup/main.go [hunk 7]
```diff
@@ -105,6 +132,22 @@ func main() {
...
+		if _, err := iamClient.DeleteOpenIDConnectProvider(ctx, &iam.DeleteOpenIDConnectProviderInput{
+			OpenIDConnectProviderArn: lo.ToPtr(names[i]),
+		}); err != nil {
+			logger.With("name", names[i]).Errorf("deleting test cluster oidc, %v", err)
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:142]:
```suggestion
			logger.With("arn", arns[i]).Errorf("deleting test cluster oidc provider, %v", err)
```

File: test/hack/cleanup/main.go [hunk 8]
```diff
@@ -105,6 +132,22 @@ func main() {
...
+		} else {
+			logger.With("name", names[i]).Infof("deleted test cluster oidc")
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:144]:
```suggestion
			logger.With("arn", names[i]).Infof("deleted test cluster oidc provider")
```

File: .github/workflows/sweeper.yaml [hunk 9]
```diff
@@ -9,14 +9,18 @@ permissions:
...
+      fail-fast: false
+      matrix:
+        region: [us-east-2, us-west-2]
```
Comment (jonathan-innis) [inline .github/workflows/sweeper.yaml:15]: Should we go ahead and add the `eu-west` region ahead of us expanding the conformance tests over there?

File: test/hack/cleanup/main.go (hunk 3 above)
Comment (jonathan-innis) [reply test/hack/cleanup/main.go:79]: What about the VPC? Can we validate the age of the owning VPC and see if this is beyond the timestamp?

File: test/hack/cleanup/main.go (hunk 4 above)
Comment (jonathan-innis) [reply test/hack/cleanup/main.go:256]: Any chance we can discover these names by introspecting the `.github/workflows` file or by reading the directory names. I have some concern that if we specify the test suites like this we are going to continually forget to update these values.

Comment (jonathan-innis) [reply test/hack/cleanup/main.go:256]: Another option is to do the inverse i.e. any OIDC provider that doens't have the `do-not-delete` tag gets cleaned up.

File: test/hack/cleanup/main.go (hunk 3 above)
Comment (engedaam) [reply test/hack/cleanup/main.go:88]: The VPC does not have creation timestamp as well

File: test/hack/cleanup/main.go (hunk 4 above)
Comment (engedaam) [reply test/hack/cleanup/main.go:272]: We don't create the OIDCs, so I'm not sure how we can tag them

File: test/hack/cleanup/main.go (hunk 5 above)
Comment (engedaam) [reply test/hack/cleanup/main.go]: For the sake of scope, we can do that when we handle resource leak monitoring

File: test/hack/cleanup/main.go [hunk 10]
```diff
@@ -68,13 +80,37 @@ func main() {
 			}
 		}
 	}
+}
 
-	// Terminate any old stacks that were provisioned as part of testing
-	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
+// Terminate any old stacks that were provisioned as part of testing
+// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
+func cleanupOldstack(ctx context.Context, ec2Client *ec2.Client, cloudWatchClient *cloudwatch.Client, cloudFormationClient *cloudformation.Client, expirationTime time.Time, logger *zap.SugaredLogger) {
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:87]: I feel like we should try to separate out deleting the security groups in the VPC from deleting the Stacks if we can. I'm not a big fan of us tying these two operations together in this one function.

File: .github/actions/e2e/cleanup/action.yaml [hunk 11]
```diff
@@ -30,6 +30,18 @@ runs:
     - uses: ./.github/actions/e2e/install-eksctl
       with:
         eksctl_version: v0.147.0
+    - name: delete-security-group
+      shell: bash
+      run: | 
+        aws ec2 describe-security-groups \
```
Comment (jonathan-innis) [inline .github/actions/e2e/cleanup/action.yaml:36]: Can we describe why we delete the security group here. Either in the YAML or in the script as a comment?

File: .github/actions/e2e/create-cluster/action.yaml [hunk 12]
```diff
@@ -125,6 +125,16 @@ runs:
         wellKnownPolicies:
           ebsCSIController: true
       EOF
+  - name: tag oidc provider of the cluster
+    shell: bash
+    run: |
+      for arn in $(aws iam list-open-id-connect-providers --query "OpenIDConnectProviderList[*].{ARN:Arn}" --output text); do
```
Comment (jonathan-innis) [inline .github/actions/e2e/create-cluster/action.yaml:131]: Nice! Can we tag it with the github run url like we do in the clusterConfig so that these tags are consistent?

File: test/cloudformation/iam_cloudformation.yaml [hunk 13]
```diff
@@ -66,12 +66,15 @@ Resources:
               - iam:GetRole
               - iam:GetRolePolicy
               - iam:ListAttachedRolePolicies
+              - iam:ListOpenIDConnectProviders
+              - iam:TagOpenIDConnectProvider
```
Comment (jonathan-innis) [inline test/cloudformation/iam_cloudformation.yaml:70]: Can we add these policies to the list of OpenID permissions? I believe there's a block that already covers OIDC permissions right now

File: test/hack/cleanup/main.go [hunk 14]
```diff
@@ -68,13 +85,42 @@ func main() {
...
+	if err != nil {
+		logger.With("error", err).Error("getting security groups")
+	}
+	logger.With("ids", ids, "count", len(ids)).Infof("discovered test security groups to delete")
+	deleted := 0
 
-	// Terminate any old stacks that were provisioned as part of testing
-	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
-	names := getOldStacks(ctx, cloudFormationClient, expirationTime)
+	workqueue.ParallelizeUntil(ctx, 10, len(ids), func(i int) {
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:98]: Is it possible for us to get throttled if we do this this way?

File: test/hack/cleanup/main.go [hunk 15]
```diff
@@ -51,10 +56,22 @@ func main() {
 	ec2Client := ec2.NewFromConfig(cfg)
 	cloudFormationClient := cloudformation.NewFromConfig(cfg)
 	cloudWatchClient := cloudwatch.NewFromConfig(cfg)
+	iamClient := iam.NewFromConfig(cfg)
 
-	// Terminate any old instances that were provisioned by Karpenter as part of testing
-	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
-	ids := getOldInstances(ctx, ec2Client, expirationTime)
+	cleanupInstances(ctx, ec2Client, cloudWatchClient, expirationTime, logger)
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:61]: When I was thinking about parallelizing, I was thinking that you would just parallelize each of these individual functions so that the categories are parallelized. I think you actually save more time that way and you're less likely to get throttled from what I could think.

File: test/suites/drift/suite_test.go [hunk 16]
```diff
@@ -171,6 +171,10 @@ var _ = Describe("Drift", Label("AWS"), func() {
 							Key:   awssdk.String("karpenter.sh/discovery"),
 							Value: awssdk.String(settings.FromContext(env.Context).ClusterName),
 						},
+						{
+							Key:   awssdk.String("creation-date"),
```
Comment (jonathan-innis) [inline test/suites/drift/suite_test.go:175]: Nice!

File: .github/actions/e2e/create-cluster/action.yaml [hunk 17]
```diff
@@ -125,6 +125,17 @@ runs:
         wellKnownPolicies:
           ebsCSIController: true
       EOF
+  - name: tag oidc provider of the cluster
```
Comment (jonathan-innis) [inline .github/actions/e2e/create-cluster/action.yaml:128]: Can we make sure that we do an `always()` on this so that we always attempt to tag the OIDC even if the previous step doesn't succeed?

Comment (jonathan-innis) [reply .github/actions/e2e/create-cluster/action.yaml:128]: That will give us a lower chance of leaking OIDC providers if something happens with the `eksctl create cluster`

File: test/cloudformation/iam_cloudformation.yaml [hunk 18]
```diff
@@ -90,6 +91,8 @@ Resources:
             Action:
               - iam:CreateOpenIDConnectProvider
               - iam:DeleteOpenIDConnectProvider
+              - iam:ListOpenIDConnectProviders
```
Comment (jonathan-innis) [inline test/cloudformation/iam_cloudformation.yaml:94]: Is there a permission that is associated with listing the OIDC provider tags since we are also doing that?

File: test/cloudformation/iam_cloudformation.yaml [hunk 19]
```diff
@@ -72,6 +72,7 @@ Resources:
               - !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/eksctl-*"
               - !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/karpenter-irsa-*"
               - !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:role/prometheus-irsa-*"
+              - !Sub "arn:${AWS::Partition}:iam::${AWS::AccountId}:oidc-provider/*"
```
Comment (jonathan-innis) [inline test/cloudformation/iam_cloudformation.yaml:75]: Why do we need to add this?

File: test/hack/cleanup/main.go [hunk 20]
```diff
@@ -35,9 +39,15 @@ const (
 
 	karpenterProvisionerNameTag = "karpenter.sh/provisioner-name"
 	karpenterLaunchTemplateTag  = "karpenter.k8s.aws/cluster"
+	karpenterSecurityGroupTag   = "karpenter.sh/discovery"
 	githubRunURLTag             = "github.com/run-url"
 )
 
+type resources interface {
+	cleanupResources(ctx context.Context)
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:47]:
```suggestion
	cleanup(ctx context.Context)
```

File: test/hack/cleanup/main.go [hunk 21]
```diff
@@ -35,9 +39,15 @@ const (
...
+	getOldResources(ctx context.Context) (ids []string, err error)
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:48]:
```suggestion
	getExpired(ctx context.Context) (ids []string, err error)
```

File: test/hack/cleanup/main.go [hunk 22]
```diff
@@ -51,59 +61,171 @@ func main() {
...
+		&stack{cloudFormationClient: cloudFormationClient, cloudWatchClient: cloudWatchClient, expirationTime: expirationTime, logger: logger},
+		&launchtemplate{ec2Client: ec2Client, cloudWatchClient: cloudWatchClient, expirationTime: expirationTime, logger: logger},
+		&oidc{iamClient: iamClient, cloudWatchClient: cloudWatchClient, expirationTime: expirationTime, logger: logger},
+	}
 
-	// Terminate any old instances that were provisioned by Karpenter as part of testing
-	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
-	ids := getOldInstances(ctx, ec2Client, expirationTime)
-	logger.With("ids", ids, "count", len(ids)).Infof("discovered test instances to delete")
+	workqueue.ParallelizeUntil(ctx, 10, len(resources), func(i int) {
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:74]: Why do we need 10 workers?

File: test/hack/cleanup/main.go (hunk 21 above)
Comment (jonathan-innis) [reply test/hack/cleanup/main.go:48]: Why does this interface even need this method if it's not being called by some external thing?

File: test/hack/cleanup/main.go [hunk 23]
```diff
@@ -51,59 +61,171 @@ func main() {
...
 	}
+}
 
-	// Terminate any old stacks that were provisioned as part of testing
-	// We execute these in serial since we will most likely get rate limited if we try to delete these too aggressively
-	names := getOldStacks(ctx, cloudFormationClient, expirationTime)
-	logger.With("names", names, "count", len(names)).Infof("discovered test stacks to delete")
+type securitygroup struct {
+	ec2Client        *ec2.Client
+	cloudWatchClient *cloudwatch.Client
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:110]: This cloudwatch client is common to all of these? Is there a way to standardize this across the interfaces so each one doesn't need this?

File: test/hack/cleanup/main.go [hunk 24]
```diff
@@ -51,59 +61,171 @@ func main() {
...
+	expirationTime   time.Time
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:111]: Same comment on expiration time? Each of these resource structs has this value so is there a good way to standardize this so it's more DRY

File: test/hack/cleanup/main.go [hunk 25]
```diff
@@ -25,19 +25,35 @@ import (
...
+	karpenterMetricNamespace = "testing.karpenter.sh/Cleanup"
 
 	karpenterProvisionerNameTag = "karpenter.sh/provisioner-name"
 	karpenterLaunchTemplateTag  = "karpenter.k8s.aws/cluster"
+	karpenterSecurityGroupTag   = "karpenter.sh/discovery"
 	githubRunURLTag             = "github.com/run-url"
 )
 
+type CleanableResourceType interface {
+	Type() string
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:47]: I still feel like we could reduce a lot of repetition if we use a `Get()` method and then just use the get method to call the logger and then fire the metric after we do the cleanup. You could even return the items that you were able to successfully cleanup and fire a metric on that.

File: test/hack/cleanup/main.go [hunk 26]
```diff
@@ -25,19 +25,35 @@ import (
...
 	"github.com/samber/lo"
+	"go.uber.org/multierr"
 	"go.uber.org/zap"
+	"k8s.io/client-go/util/workqueue"
 )
 
 const (
 	expirationTTL            = time.Hour * 12
-	karpenterMetricNamespace = "testing.karpenter.sh/cleanup"
+	karpenterMetricNamespace = "testing.karpenter.sh/Cleanup"
```
Comment (jonathan-innis) [inline test/hack/cleanup/main.go:38]: Why the change here?

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

**Review by jonathan-innis — APPROVED** [review]
Awesome work 🚀
